62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+        for (uint32_t bezierIdx = 0; bezierIdx < beziers.size(); bezierIdx++)
+        {
+            auto hatchBezier = &beziers[bezierIdx];
+            Segment segment;
+            segment.originalBezier = hatchBezier;
+            segment.t_start = 0.0;
+            segment.t_end = 1.0;
+            segments.push_back(segment);
+        }
+
+        // TODO better way to do this
+        std::sort(segments.begin(), segments.end(), [&](Segment a, Segment b) { return a.originalBezier->p[0][major] > b.originalBezier->p[0][major]; });
+		for (Segment& segment : segments)
+		{
+			starts.push(segment);
+			//std::cout << "Starts: " << segment.originalBezier->p[0][major]  << "\n";
+		}
+
+        std::sort(segments.begin(), segments.end(), [&](Segment a, Segment b) { return a.originalBezier->p[2][major] > b.originalBezier->p[2][major]; });
+        for (Segment& segment : segments)
+		{
+			ends.push(segment.originalBezier->p[2][major]);
+			//std::cout << "Ends: " << segment.originalBezier->p[2][major] << "\n";
+		}
+        maxMajor = segments.front().originalBezier->p[2][major];
+    }
+
+    // Sweep line algorithm
+	std::priority_queue<double, std::vector<double>, std::greater<double> > intersections; // Next intersection points as major coordinate
+    std::vector<Segment> activeCandidates; // Set of active candidates for neighbor search in sweep line
+
+	// if we weren't spawning quads, we could just have unsorted `vector<Bezier*>`
+	auto candidateComparator = [&](const Segment& lhs, const Segment& rhs)
+	{
+		// btw you probably want the beziers in Quadratic At^2+B+C form, not control points
+		double _lhs = lhs.originalBezier->evaluateBezier(lhs.t_start)[minor];
+		double _rhs = rhs.originalBezier->evaluateBezier(rhs.t_start)[minor];
+
+		auto len = [](float64_t2 vec)
+		{
+			return (double)sqrt(vec.x * vec.x + vec.y * vec.y);
+		};
+		double lenLhs = len(lhs.originalBezier->p[2] - lhs.originalBezier->p[0]);
+		double lenRhs = len(rhs.originalBezier->p[2] - rhs.originalBezier->p[0]);
+		auto minLen = std::min(lenLhs, lenRhs);
+
+		//printf("Comparing bezier lhs = (%f, %f)..(%f, %f) rhs = (%f, %f)..(%f, %f) || minor at t_start (%f, %f) = %f < %f || scale of the curves: lhs = %f rhs = %f minLen = %f || abs(_lhs - _rhs) / minLen = %f\n",
+		//	lhs.originalBezier->p[0].x, lhs.originalBezier->p[0].y,
+		//	lhs.originalBezier->p[2].x, lhs.originalBezier->p[2].y,
+		//	rhs.originalBezier->p[0].x, rhs.originalBezier->p[0].y,
+		//	rhs.originalBezier->p[2].x, rhs.originalBezier->p[2].y,
+		//	lhs.t_start, rhs.t_start,
+		//	_lhs, _rhs,
+		//	len(lhs.originalBezier->p[2] - lhs.originalBezier->p[0]),
+		//	len(rhs.originalBezier->p[2] - rhs.originalBezier->p[0]),
+		//	minLen,
+		//	abs(_lhs - _rhs) / minLen
+		//);
+		// Threshhold here for intersection points, where the minor values for the curves are
+		// very close but could be smaller, causing the curves to be in the wrong order
+		// TODO: figure out if this is the best ay to do this
```

Comment:
yea that's fine to do.

tip : we can get avoid divisions in comparisons so change to `abs(_lhs - _rhs) < 1e-3 * minLen`

I'd test it a bit more with the threshold, try to strategically place bezier end/start points near intersection points with different tangetns to mess with it and see if this fails -> the tighter (closer to 0) the threshold the better
btw make the threshold a `constexpr float64_t MinorPositionComparisonThreshold` and add a comment 
`// this threhsold is used to decide when to consider minor position to be the same and check tangents because intersection algorithms has rounding errors`